_posts/2021-04-18-api-conversion.md

```diff
@@ -0,0 +1,89 @@
+---
+layout: post
+title:  "API versions conversion"
+date: 2021-04-18
+author: jakub_stejskal
+---
+Upgrade process of Strimzi and Kafka is not trivial.
+In Strimzi 0.22 we introduced new API version `v1beta2` which will replace `v1beta1` / `v1alpha1` APIs and make Strimzi ready for Kubernetes 1.22.
+This blog post will show you how to perform the conversion and get ready for next Strimzi releases.
+
+<!--more-->
+
+Even with these changes, there are still some disadvantages:
+* For upgrade, roll out the new brokers while first using the older `log.message.format.version` or `inter.broker.protocol.version` and only afterwards change to the new versions for message format and inter-broker protocol.
+* Downgrade won't be executed when new `log.message.format.version` or `inter.broker.protocol.version` are already used.
+  You need to change configuration to use same `log.message.format.version` and `inter.broker.protocol.version` as older Kafka.
+
+You can find further information in our [documentation](https://strimzi.io/docs/operators/0.22.1/full/deploying.html#proc-upgrade-cli-tool-crds-str).
+
+### New API versions
+
+As we stated before, we introduced new API version `v1beta2` for Strimzi resources.
+Every user of Strimzi will have to the migration from old `v1beta1` to `v1beta2`.
+This migration is not necessarily needed right after upgrade to Strimzi 0.22, but has to be done before upgrade to Strimzi 0.23 or later.
+The main reason why Strimzi users had to do this migration are preparations of Strimzi to use [Kubernetes CRD `apiextensions/v1`](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#customresourcedefinition-v122).
+
+Custom resources are edited and controlled using APIs added to Kubernetes by CRDs.
+Put another way, CRDs extend the Kubernetes API to allow the creation of custom resources.
+CRDs are themselves resources within Kubernetes.
+They are installed in a Kubernetes cluster to define the versions of API for the custom resource.
+Each version of the custom resource API can define its own schema for that version.
+Kubernetes clients, including the Strimzi Operators, access the custom resources served by the Kubernetes API server using a URL path (API path), which includes the API version
+
+In 0.21, we supported custom resource version `v1alpha1` for all custom resources and `v1beta1` for `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker`, `KafkaUser` and `KafkaTopic`.
+These versions are currently deprecated and will be dropped in the following Strimzi 0.23.
+
+One another change is, that with new API version we remove some deprecated fields so users need to migrate them to new format.
+For instance `v1beta2` supports only new array based listener configuration and new metrics configuration formats from config maps.
+In additional some old unused fields like `tlsSidecar` were removed as well.
```

Comment:
might be worth mentioning also the metrics or logging changes which are more important than `tlsSidecars`